src/main/kotlin/io/hsar/wh40k/combatsimulator/CombatSimulation.kt

```diff
@@ -18,16 +21,51 @@ class CombatSimulation(val world: World) {
             .sortedByDescending { (initiativeRoll, _) -> initiativeRoll }
             .map { (_, eachUnit) -> eachUnit }
 
-    fun run() {
+    fun runRound() {
         initiativeOrder.forEach { unit ->
-            unit.tacticalActionStrategy
-                    .decideTurnActions(world, unit, unit.availableActionOptions)
-                    .let { actionsToExecute ->
-                       unit.executeActions(actionsToExecute)
-                    }
-                    .also {
-                        world.removeUnits(world.findDead())
-                    }
+            if((unit.currentAttributes.getValue(Attribute.CURRENT_HEALTH) as NumericValue).value > 0) {
+                // check if unit alive rather than try removing dead units from list as concurrency risks
+                // caller needs to exception handle this in case CURRENT_HEALTH not provided in json
+                unit.tacticalActionStrategy
+                        .decideTurnActions(world, unit, unit.availableActionOptions)
+                        .let { actionsToExecute ->
+                            unit.executeActions(actionsToExecute)
+                        }
+                        .also {
+                            val deadUnits = world.findDead()
+                            deadUnits.forEach { deadUnit->
+                                println("${deadUnit.name} died")
+                            }
+                            world.removeUnits(deadUnits)
+                        }
+            }
+
         }
     }
+
+    fun runSimulation() {
+        var roundNum = 1
+        while(world.enemyForces.size > 0 && world.friendlyForces.size > 0) {
+            printRoundStartStatus(roundNum)
+            try {
+                runRound()
+            } catch(e: Exception) {
```

Comment:
I appreciate the desire for safety but I don't agree with this, I think that our system is fairly closed and controlled and thus we should never see exceptions in normal operation (this is the correct use of exceptions for "exceptional behaviour").

That's a very roundabout way of saying that we should just let exceptions reach the user and terminate the run. It's not like it's super expensive to start another run.

Reply:
Removed.